File: Notes/code/binary_index_tree.cpp
int sums[N + 1];

void update(int pos, int val){
	for(; pos <= N; pos += pos & -pos)
		sums[pos] += val;
}

int query(int pos){
	int sum = 0;
	for(; pos > 0; pos -= pos & -pos)
		sum += sums[pos];
	return sum;
}

int rangeQuery(int start, int end){
	return query(end) - query(start - 1);
}

File: Notes/code/binary_index_tree.py
class BinaryIndexTree:

  def __init__(self, N):
    self.N = N
    self.tree = [0] * (N + 1)

  def update(self, pos, val):
    while pos <= self.N:
      self.tree[pos] += val
      pos += pos & -pos

  def query(self, pos):
    total = 0
    while pos > 0:
      total += self.tree[pos]
      pos -= pos & -pos
    return total

  def rangeQuery(self, start, end):
    return self.query(end) - self.query(start - 1)
